### PR TITLE
New Plugin: Schrack i-CHARGE CION

### PR DESCRIPTION
**Pull request by mzanetti**
nymea-plugins-modbus pull request checklist:

- [x] Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin name"

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [x] Did you update translations (`cd builddir && make lupdate`)?

- [ ] If you added a new plugin, should it be added to nymea-plugins-all or nymea-plugins-maker?

### CODE REVIEW

File: schrack/cion-registers.json [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        },
+        {
+            "id": "pluggedInDuration",
+            "address": 153,
+            "size": 2,
+            "type": "uint32",
+            "registerType": "holdingRegister",
+            "readSchedule": "update",
+            "description": "Plugged in duration",
+            "defaultValue": 6,
```
Comment (t-mon) [inline schrack/cion-registers.json:76]: I find it useful to set also the unit as string coming from the register, so when you look the next time to the code or you are about to use this property you know right away which unit the register provides from the comments.

File: schrack/foobar.cpp [hunk 2]
```diff
@@ -0,0 +1,536 @@
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
```
Comment (t-mon) [inline schrack/foobar.cpp:1]: Is `foobar` class a leftover from playing with the tool or do you need those somewhere?

File: schrack/schrack.pro [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    integrationpluginschrack.h \
+    cionmodbusrtuconnection.h \
+    ../modbus/modbusdatautils.h
+
+OTHER_FILES += cion-registers.json
+
+modbusconnection.commands = python $${top_srcdir}/modbus/tools/generate-connection.py -j $${_PRO_FILE_PWD_}/cion-registers.json -o $${_PRO_FILE_PWD_} -c CionModbusRtuConnection
+QMAKE_EXTRA_TARGETS += modbusconnection
+
+#target.depends += modbusconnection
```
Comment (t-mon) [inline schrack/schrack.pro:20]: Nice, I would not auto generate the classes on every build, so commenting this out is good. The only reason why I don't wanna generate the classes during build is that the tools changes over time, and you might want to migration of the new class individually. Once the tool is quiet stable we might wanna consider that.

**Review by t-mon — CHANGES_REQUESTED** [review]

File: schrack/schrack.pro (hunk 3 above)
Comment (mzanetti) [reply schrack/schrack.pro:20]: Yes... that's why I removed it. Also I ran into another problem with autogenerating which would need to be solved first but I can't recall atm what it was. Probably build-dependencies to python and such...

In any case, having the option to `make modbusconnection` to update everything is nice, I've been using it all the time. Eventually we should move that to the .pri so all modbus plugins have it available right away

File: schrack/cion-registers.json (hunk 1 above)
Comment (mzanetti) [reply schrack/cion-registers.json:92]: done

File: schrack/schrack.pro (hunk 3 above)
Comment (t-mon) [reply schrack/schrack.pro:20]: Right, so the plan is any way to move the modbus folder into a lib, and then also move the tool to a own dev package.

**Review by t-mon — APPROVED** [review]

**Comment by mzanetti** [comment]
```cionmodbusrtuconnection.cpp:237:96: error: ‘pow’ was not declared in this scope
  237 |                     float receivedU1Voltage = ModbusDataUtils::convertToUInt16(values) * 1.0 * pow(10, -2);
      |                                                                                                ^~~
cionmodbusrtuconnection.cpp: In lambda function:
cionmodbusrtuconnection.cpp:266:98: error: ‘pow’ was not declared in this scope
  266 |                     float receivedGridVoltage = ModbusDataUtils::convertToUInt16(values) * 1.0 * pow(10, -2);
      |                                                                                                  ^~~
make[3]: *** [Makefile:445: cionmodbusrtuconnection.o] Error 1
```

the modbus tool probably is lacking an include and should perhaps use qPow() from QMath instead

**Comment by t-mon** [comment]
> the modbus tool probably is lacking an include and should perhaps use qPow() from QMath instead

This has already been fixed in the new tool improvement branch https://github.com/nymea/nymea-plugins-modbus/pull/62

**Pull request merged**